Created an application for testing students for lesson 1

KorolkovRS opened the pull request:
# Global tasks
- Create an application for testing students using Spring IoC.
# Сhanges
- Created controller, service and dao layers.
- Spring context is described by an `spring-context.xml`.
- `ResourceProviderServiceImpl` covered by unit-tests.
- The questions for testing are put in `csv\questions.csv`.
# Completed tasks
- [X] All classes in the application must solve a strictly defined task.
- [X] The context is described by an XML file.
- [X] All dependencies must be configured in the IoC container.
- [X] The name of the resource with questions (CSV file) must be encoded with a line in an XML file with context.
- [X] CSV with questions is read exactly as a resource, not as a file.
- [X] You don't need to put Scanner, PrintStream and other standard types in the context!
- [X] All I/O is done in English.
- [X] Write a unit test of some service.
- [X] The application must be launched correctly using "java -jar"

KorolkovRS committed:
Intermediate commit. Spring project is configured. Add dependencies in pom.xml

KorolkovRS committed:
Intermediate commit. Added dao, service and controller layers.

KorolkovRS committed:
Intermediate commit. ResourceProviderServiceImpl is covered by tests.

KorolkovRS committed:
Intermediate commit. Update pom.xml and ResourceProviderServiceImplTest closes the resources behind it

KorolkovRS committed:
Intermediate commit. Fixed an error with unavailable resources from the jar archive

KorolkovRS committed:
Intermediate commit. Code refactoring.

KorolkovRS committed:
Add README.md

stvort commented on spring-01/src/main/java/ru/korolkovrs/spring/controller/QuestionControllerImpl.java:
```diff
@@ -0,0 +1,21 @@
+    public void getAll() {
+        List<Question> questions = dao.getAll();
+        for (Question q: questions) {
+            System.out.println(q);
+            System.out.println("**************************");
```
Работу с консолью (ввод/вывод, System.in/out) можно вынести в отдельный сервис. Так будет удобнее тестировать классы, где этот ввод/вывод используется. Причем System.in/out такому сервису можно будет тоже передать через конструктор. Такой сервис ничего не будет знать о предметной области и может быть переиспользован

stvort commented on spring-01/src/main/java/ru/korolkovrs/spring/controller/QuestionControllerImpl.java:
```diff
@@ -0,0 +1,21 @@
+import java.util.List;
+
+@RequiredArgsConstructor
+public class QuestionControllerImpl implements QuestionController {
```
Не уверен, что это контроллер. Скорее просто сервис

stvort commented on spring-01/src/main/java/ru/korolkovrs/spring/controller/QuestionControllerImpl.java:
```diff
@@ -0,0 +1,21 @@
+    private final QuestionDao dao;
+
+    @Override
+    public void getAll() {
```
Метод, название которого начинается на `get` желательно чтобы что-то возвращал. Возможно лучше `printAll`

stvort commented on spring-01/src/main/java/ru/korolkovrs/spring/dao/QuestionDaoImpl.java:
```diff
@@ -0,0 +1,27 @@
+    public List<Question> getAll() {
+        List<Question> beans = null;
+        try(Reader in = resourceDiscoverService.getResourceReader()) {
+            beans = new CsvToBeanBuilder(in)
```
`new CsvToBeanBuilder<Question>`

stvort commented on spring-01/src/main/java/ru/korolkovrs/spring/dao/QuestionDaoImpl.java:
```diff
@@ -0,0 +1,27 @@
+            beans = new CsvToBeanBuilder(in)
+                    .withType(Question.class).build().parse();
+        } catch (IOException e) {
+            e.printStackTrace();
```
Исключения лучше не подавлять, а заворачивать в свои бизнес исключения (например, `QuestionsLoadingException`). Ну и возможно лучше, чтобы это были unchecked исключения

KorolkovRS replied:
IOException выбрасывается при ошибке закрытия ресурса в ` try(Reader in = resourceDiscoverService.getResourceReader())`. Ошибки чтения ресурса IllegalStateException все еще пробрасываются наружу.
Не уверен, что стоит выбрасывать unchecked исключение при ошибке закрытия ресурса.

stvort commented on spring-01/src/main/java/ru/korolkovrs/spring/domain/Question.java:
```diff
@@ -0,0 +1,35 @@
+    private List<String> answers;
+
+    @Override
+    public String toString() {
+        StringBuilder sb = new StringBuilder(String.format("Question №%d: %s\nAnswers:\n", id, question));
+        if (answers != null) {
+            for (int i = 0; i < answers.size(); i++) {
+                sb.append(i + 1 + ") " + answers.get(i) + "\n");
+            }
+        }
+        sb.deleteCharAt(sb.length() - 1);
+        return sb.toString();
+    }
```
Лучше не использовать `toString` для вывода информации пользователю. Доменный класс один, а желающих его вывести может быть много. Так что лучше сделать сервис, который переводит объекты в строку

stvort commented on spring-01/src/main/java/ru/korolkovrs/spring/domain/Question.java:
```diff
@@ -0,0 +1,35 @@
+    private String question;
+
+    @CsvBindAndSplitByName(column = "answer", elementType = String.class, splitOn = "\\|", collectionType =  ArrayList.class)
+    private List<String> answers;
```
Как узнать, какой из ответов правильный? Возможно лучше `List<Answer>`

stvort commented on spring-01/src/main/java/ru/korolkovrs/spring/service/ResourceProviderService.java:
```diff
@@ -0,0 +1,7 @@
+
+import java.io.Reader;
+
+public interface ResourceProviderService {
```
Эта штука используется в дао, а дао ниже по слою нежели сервисы и не может их использовать. Лучше убрать `Service` из названия

stvort commented on spring-01/src/main/java/ru/korolkovrs/spring/service/ResourceProviderService.java:
```diff
@@ -0,0 +1,7 @@
+import java.io.Reader;
+
+public interface ResourceProviderService {
+    Reader getResourceReader();
```
Я бы просто поток отдавал, а уж вызывающий код решил бы, чем его читать

stvort commented on spring-01/src/main/java/ru/korolkovrs/spring/service/ResourceProviderServiceImpl.java:
```diff
@@ -0,0 +1,21 @@
+    @Override
+    public Reader getResourceReader() {
+        ClassLoader classLoader = getClass().getClassLoader();
+        Reader reader = new InputStreamReader(classLoader.getResourceAsStream(fileName));
+        if (reader == null) {
+            throw new RuntimeException("File not found! " + fileName);
+        }
```
Ридер не будет `null` т.к. он выкинет исключение если передаваемый в него поток равен `null`

stvort commented on spring-01/src/test/java/ru/korolkovrs/spring/service/ResourceProviderServiceImplTest.java:
```diff
@@ -0,0 +1,38 @@
+            assertThat(reader)
+                    .isNotNull()
+                    .isInstanceOf(Reader.class);
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
```
Зачем нужен `catch`? Так тест, который мог упасть, может пройти

KorolkovRS replied:
Catch обрабатывает IOExctption, выбрасываемый при закрытии потока. К тестированию метода он не относится.

KorolkovRS committed:
Refactoring: 

- QuestionController class has been renamed to QuestionService
- getAll() method has been renamed to printAll() and update 
- toString() has been removed from Question class 
- Add QuestionToStringConverter class 
- getResourceReader() method has been renamed to getResourceStream(). Return type change to InputStream
